lib/lobby.ts

```diff
@@ -0,0 +1,184 @@
+}
+
+/**
+ * The class containing a user and their data
+ */
+export class Lobby extends DbItem implements ILobby {
+  /**
+   * A static function to query for a user from their id
+   *
+   * @returns a user object if the id exists in the database
+   */
+  public static async fromId(id: string): Promise<Lobby | null> {
+    const client = await getClient();
+    const document = await client.findDbItem(COLLECTION.LOBBIES, id);
+    if (!document) return null;
+    const content: DatabaseEntry = document.getContent() as DatabaseEntry;
+    return new Lobby(id, content, document.key ?? null);
+  }
+
+  /**
+   * A static function to create a new Lobby and corresponding Spotify playlist
+   *
+   * @returns a newly created Lobby object
+   */
+  public static async create(props: LobbyProps, key : string | null = null) : Promise<Lobby | null> {
+    // create new spotify playlist
+    const playlistId = await createSpotifyPlaylist();
+    const newProps = {...props, spotifyPlaylistId: playlistId};
+    return new Lobby(playlistId, newProps, key);
+  }
+
+  /**
+   * The spotify id of playlist
+   */
+  public readonly spotifyPlaylistId?: string;
+
+  /**
+   * The manager user id of a lobby
+   */
+  public readonly managerId: string;
+
+  /**
+   * The theme of a lobby
+   */
+  public readonly theme: string;
+
+  /**
+   * The playlist name
+   */
+  get name(): string {
+    return this.#name;
+  }
+  #name: string;
+
+  /**
+   * The participants in a lobby
+   */
+  get participants(): string[] {
+    return this.#participants;
+  }
+
+  #participants: string[];
+
+  /**
+   * The list of song ids in the playlist
+   */
+  get songIds(): string[] {
+    return this.#songIds;
+  }
+
+  #songIds: string[];
+
+  protected constructor(id: string, props: LobbyProps, key: string | null = null) {
+    super(id, COLLECTION.LOBBIES, key);
+    this.managerId = props.managerId;
+    this.#participants = props.participants ?? [];
+    this.#songIds = props.songIds ?? [];
+    this.spotifyPlaylistId = props.spotifyPlaylistId;
+    this.theme = props.theme;
+    this.#name = props.name;
+  }
+
+  /**
+   * Converts the object into a form for the database
+   * @returns a database entry
+   */
+  public toJson(): DatabaseEntry {
+    const {collectionName: _c, ...entry} = this;
+    return { ...entry, name: this.#name, participants: this.#participants };
+  }
+
+  /**
+   * validate that a calling user is the manager of a lobby
+   */
+  public async validateManagerAccess(user: User): Promise<boolean> {
+    return this.managerId === user.id;
+  }
+
+  /**
+   * Add a user to the lobby
+   */
+  public async addUser(user: User, addUserId: string, writeToDatabase = true): Promise<boolean> {
+    const validate = await this.validateManagerAccess(user);
+    if (validate) {
+      this.#participants = [...this.#participants, addUserId];
+      writeToDatabase && void this.writeToDatabase();
+      return true;
+    }
+    return false;
```

Comment:
nit: while this order is fine, you will see the following more in industry
```suggestion
    if (!validate) {
      return false;
    }
    this.#participants.push(addUserId);
    writeToDatabase && void this.writeToDatabase();
    return true;
```

usually, you want your main logic outside a branch statement. why? your main logic is usually the biggest chunk of code in your function. In terms of readability, it is usually better to have **less** indents or nested logic.

By moving fail conditions up, you reduce the amount of nesting and let the reader quickly discern the fail conditions for the function

Reply:
Makes sense!